File: maximum-sum-circular-subarray/maximum-sum-circular-subarray.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int currMax = 0;
        int currMin = 0;
        int maxSum = nums[0];
        int minSum = nums[0];
        int total = 0;

        for (int i = 0; i < nums.length; i++) {
            currMax = Math.max(currMax + nums[i], nums[i]);
            maxSum = Math.max(currMax, maxSum);

            currMin = Math.min(currMin + nums[i], nums[i]);
            minSum = Math.min(currMin, minSum);

            total += nums[i];
        }

        if (maxSum < 0) return maxSum;
        
        return Math.max(maxSum, total - minSum);
    }
}

File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        currMax = 0
        currMin = 0
        bestMax = nums[0]
        bestMin = nums[0]
        total = 0
        for num in nums:
            currMax = max(currMax + num, num)
            currMin = min(currMin + num, num)
            bestMax = max(bestMax, currMax)
            bestMin = min(bestMin, currMin)
            total += num
        if total == bestMin:
            return bestMax
        return max(bestMax, total - bestMin)